bmap-rs/src/main.rs

```diff
@@ -75,49 +83,117 @@ impl SeekForward for Decoder {
     }
 }
 
-fn setup_input(path: &Path) -> Result<Decoder> {
+async fn fetch_url_http(url: hyper::Uri) -> Result<Response<Body>, hyper::Error> {
+    let client = Client::new();
+    client.get(url.clone()).await
+}
+async fn setup_remote_input(url: Uri, fpath: &Path) -> Result<Input> {
+    if fpath.extension().unwrap() != "gz" {
+        bail!("Image file format not implemented")
+    }
+    let input = Input::Remote(match url.scheme_str() {
+        Some("https") =>panic!("This feature doesn't work because it needs implementing async enviroment and fetch_url_https function"),
+        Some("http") => fetch_url_http(url).await?,
+        _ => {
+            bail!("url is not http or https")
+        }
+    });
+    Ok(input)
+}
+
+fn setup_local_input(path: &Path) -> Result<Input> {
     let f = File::open(path)?;
     match path.extension().and_then(OsStr::to_str) {
         Some("gz") => {
             let gz = GzDecoder::new(f);
-            Ok(Decoder::new(Discarder::new(gz)))
+            Ok(Input::Local(Decoder::new(Discarder::new(gz))))
         }
-        _ => Ok(Decoder::new(f)),
+        _ => Ok(Input::Local(Decoder::new(f))),
     }
 }
 
-fn copy(c: Copy) -> Result<()> {
-    if !c.image.exists() {
-        bail!("Image file doesn't exist")
-    }
+async fn copy(c: Copy) -> Result<()> {
+    let url = c
+        .image
+        .to_str()
+        .expect("Fail to convert remote path to str")
+        .parse::<Uri>()
+        .expect("Fail to convert remote path to url");
+    let input_type = match url.scheme() {
+        Some(_) => setup_remote_input(url.clone(), &c.image).await?,
+        None => {
+            if !c.image.exists() {
+                bail!("Image file doesn't exist")
+            }
+            setup_local_input(&c.image)?
+        }
+    };
+
+    let bmap = match input_type {
+        Input::Local(_) => find_bmap(&c.image).ok_or_else(|| anyhow!("Couldn't find bmap file"))?,
+        Input::Remote(_) => {
+            let img_path = url.path();
+            find_bmap(Path::new(&img_path[6..]))
+                .ok_or_else(|| anyhow!("Couldn't find bmap file"))?
```

Comment:
also this is quite awkward as we need to download the bmap file manually; would it be possible to download the bmap file from the remote source too ?